### PR TITLE
[Bitwarden] Add logout action and handle CLI errors, empty password, loading states, API version

### PR DESCRIPTION
**Pull request by daniel-stoneuk**
## Description

* Added lock symbol
* Added logout vault action
* Added error handling to the refresh action (may fix https://github.com/raycast/extensions/issues/1285)
* Handle empty password
* Improve loading and empty vault messages
* Update Raycast API and add tooltip

## Screenshots

Screenshot 2022-04-05 at 21 29 38

Screenshot 2022-04-06 at 13 02 50

Screenshot 2022-04-06 at 13 03 04

Screenshot 2022-04-06 at 13 03 35

Screenshot 2022-04-06 at 13 03 45

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed in the `metadata` folder

### CODE REVIEW

File: extensions/bitwarden/src/search.tsx [hunk 1]
```diff
@@ -97,12 +97,33 @@ export function ItemList(props: { api: Bitwarden }) {
...
-      await loadItems(session.token);
-      await toast.hide();
+      try {
+        await bitwardenApi.sync(session.token);
+        await loadItems(session.token);
+        await toast.hide();
+      } catch (error) {
+        await bitwardenApi.logout();
+        await session.deleteToken();
+        toast.style = Toast.Style.Failure;
```
Comment (pomdtr) [inline extensions/bitwarden/src/search.tsx:107]: Will this update the toast ? I did not know it was possible.

File: extensions/bitwarden/src/search.tsx [hunk 2]
```diff
@@ -143,13 +160,28 @@ export function ItemList(props: { api: Bitwarden }) {
...
-                <Action icon={Icon.ArrowClockwise} title={"Refresh Items"} onAction={refreshItems} />
-              </ActionPanel>
+              !state.isLoading && (
+                <ActionPanel>
+                  <VaultActions refreshItems={refreshItems} lockVault={lockVault} logoutVault={logoutVault} />
+                </ActionPanel>
+              )
             }
           />
         </Fragment>
```
Comment (pomdtr) [inline extensions/bitwarden/src/search.tsx:170]: I think the fragment is no longer necessary here.

**Review by pomdtr — COMMENTED** [review]
Looks good to me.

Can you add a toast when the user is logging in the unlock form ? It would allow the user to understand what is the point of logout vs locking the vault.

File: extensions/bitwarden/src/search.tsx (hunk 1 above)
Comment (daniel-stoneuk) [reply extensions/bitwarden/src/search.tsx:107]: Yep!

**Comment by daniel-stoneuk** [comment]
Cheers, I was unsure about exposing the `logout` action since to the user there is no change to a logged out vault. I'll add a form description field to the login page with the current state.

**Comment by daniel-stoneuk** [comment]
Screenshots of the `UnlockForm`

### Brief loading state:

Screenshot 2022-04-05 at 21 29 18

### Logged out:
Screenshot 2022-04-05 at 21 29 20

### Locked:
Screenshot 2022-04-05 at 21 29 38

File: extensions/bitwarden/src/search.tsx (hunk 2 above)
Comment (daniel-stoneuk) [reply extensions/bitwarden/src/search.tsx:170]: The empty state is still in the same part of the tenary statement here I think

Comment (pomdtr) [reply extensions/bitwarden/src/search.tsx:170]: I don't think we need the ternary statement here. The List.EmptyView is supposed to be next to the list items. I agree that the API is confusing.

The <Fragment> is also unecessary.

**Comment by pomdtr** [comment]
The Vault Status field is a good idea !

I saw an error I made while doing your review. When the user is logged out, the UnlockForm component calls `bw login` then `bw logout --raw`. This is unessary, since `bw login --raw` already return the sessionToken. Could you correct this in your pr ?

File: extensions/bitwarden/src/search.tsx (hunk 2 above)
Comment (daniel-stoneuk) [reply extensions/bitwarden/src/search.tsx:170]: I did this for when the list of items is empty so a different empty message could be shown. Initially I was going to put different actions in the empty states which is why I kept it (and hence the Fragment was required in the true clause).

I'll move the `if` out of the return in my next commit! Cheers.

**Comment by daniel-stoneuk** [comment]
Thanks, glad you like it. 

I'm not sure what you mean regarding the UnlockForm and session token, please could you clarify?

**Comment by daniel-stoneuk** [comment]
Ah, just had a re-read. I'll remove the second `bw unlock --raw` after `bw login --raw`.

**Comment by daniel-stoneuk** [comment]
Should we make the wording "Refresh" and "Syncing" consistent? I'm thinking changing the refresh action to "Sync Vault" to match the CLI.

**Comment by pomdtr** [comment]
> Should we make the wording "Refresh" and "Syncing" consistent? I'm thinking changing the refresh action to "Sync Vault" to match the CLI.

Sounds good to me !

**Comment by daniel-stoneuk** [comment]
Great!

It seems like the sync action is currently affected by https://github.com/bitwarden/jslib/pull/749, which means we show an empty vault on sync (after a while). Since we include the suggestion to try logging out and a fix should be included in the next release, I think we don't need to add any workaround / explicit note in the UI.

**Pull request merged**